staker/l1_validator.go

```diff
@@ -271,68 +232,90 @@ func (v *L1Validator) generateNodeAction(ctx context.Context, stakerInfo *OurSta
 		return nil, false, nil
 	}
 
-	startBlock := v.l2Blockchain.GetBlockByHash(startState.GlobalState.BlockHash)
-	if startBlock == nil && (startState.GlobalState != validator.GoGlobalState{}) {
-		expectedBlockHeight, inboxPositionInvalid, err := v.blockNumberFromGlobalState(startState.GlobalState)
-		if err != nil {
-			return nil, false, fmt.Errorf("error getting block number from global state: %w", err)
+	caughtUp, startCount, err := GlobalStateToMsgCount(v.inboxTracker, v.txStreamer, startState.GlobalState)
+	if err != nil {
+		return nil, false, err
+	}
+	if !caughtUp {
+		target := GlobalStatePosition{
+			BatchNumber: startState.GlobalState.Batch,
+			PosInBatch:  startState.GlobalState.PosInBatch,
 		}
-		if inboxPositionInvalid {
-			log.Error("invalid start global state inbox position", startState.GlobalState.BlockHash, "batch", startState.GlobalState.Batch, "pos", startState.GlobalState.PosInBatch)
-			return nil, false, errors.New("invalid start global state inbox position")
+		var current GlobalStatePosition
+		head, err := v.txStreamer.GetProcessedMessageCount()
+		if err != nil {
+			_, current, err = v.blockValidator.GlobalStatePositionsAtCount(head)
 		}
-		latestHeader := v.l2Blockchain.CurrentBlock()
-		if latestHeader.Number.Int64() < expectedBlockHeight {
-			log.Info("catching up to chain blocks", "localBlocks", latestHeader.Number, "target", expectedBlockHeight)
-			return nil, false, nil
+		if err != nil {
+			log.Info("catching up to chain messages", "target", target)
+		} else {
+			log.Info("catching up to chain blocks", "target", target, "current", current)
 		}
-		log.Error("unknown start block hash", "hash", startState.GlobalState.BlockHash, "batch", startState.GlobalState.Batch, "pos", startState.GlobalState.PosInBatch)
-		return nil, false, errors.New("unknown start block hash")
+		return nil, false, nil
 	}
 
-	var lastBlockValidated uint64
+	var validatedCount arbutil.MessageIndex
+	var validatedGlobalState validator.GoGlobalState
 	if v.blockValidator != nil {
-		var expectedHash common.Hash
-		var validRoots []common.Hash
-		lastBlockValidated, expectedHash, validRoots = v.blockValidator.LastBlockValidatedAndHash()
-		haveHash := v.l2Blockchain.GetCanonicalHash(lastBlockValidated)
-		if haveHash != expectedHash {
-			return nil, false, fmt.Errorf("block validator validated block %v as hash %v but blockchain has hash %v", lastBlockValidated, expectedHash, haveHash)
+		valInfo, err := v.blockValidator.ReadLastValidatedInfo()
+		if err != nil {
+			return nil, false, err
+		}
+		validatedGlobalState = valInfo.GlobalState
+		caughtUp, validatedCount, err = GlobalStateToMsgCount(v.inboxTracker, v.txStreamer, valInfo.GlobalState)
+		if err != nil {
+			return nil, false, fmt.Errorf("%w: not found validated block in blockchain", err)
+		}
+		if !caughtUp {
+			log.Info("catching up to last validated block", "target", valInfo.GlobalState)
```

Comment:
This is expected when recovering from an unorderly shutdown.
I think not in correct code, but with reasonable changes, it could also happen while syncing a new node (treating accepted assertion as validated)